internal/liqonet/route-operator/symmetricRoutingOperator.go

```diff
@@ -0,0 +1,165 @@
+// the local CNI but if the reverse path filtering is in strict mode (default) the traffic coming from a
+// network interface where the source address is not routable by the same interface is dropped. To solve the
+// issue it requires to set the reverse path filtering to loose mode, but some times it is not possible. To
+// overcome the problem all the traffic sent to or coming from a peering cluster has to be routed using the
+// overlay network.
+func (src *SymmetricRoutingController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var p corev1.Pod
+	var err error
+	err = src.Get(ctx, req.NamespacedName, &p)
+	if err != nil && !k8sApiErrors.IsNotFound(err) {
+		klog.Errorf("an error occurred while getting pod {%s}: %v", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+	if k8sApiErrors.IsNotFound(err) {
+		// Remove the peer.
+		deleted, err := src.delRoute(req)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		if deleted {
+			klog.Infof("successfully removed route for pod {%s} from vxlan overlay network", req.String())
+		}
+		return ctrl.Result{}, nil
+	}
+	added, err := src.addRoute(req, &p)
+	if err != nil {
+		klog.Errorf("an error occurred while adding route for pod {%s} with IP address {{%s}} to the vxlan overlay network: %v",
+			req.String(), p.Status.PodIP, err)
+		return ctrl.Result{}, err
+	}
+	if added {
+		klog.Errorf("successfully added route for pod {%s} with IP address {%s} to the vxlan overlay network: %v",
+			req.String(), p.Status.PodIP)
+	}
+	return ctrl.Result{}, nil
+}
+
+// NewSymmetricRoutingOperator returns a new controller ready to be setup and started with the controller manager.
+func NewSymmetricRoutingOperator(nodeName string, routingTableID int, vxlanDevice overlay.VxlanDevice,
+	nodesLock *sync.RWMutex, vxlanNodes map[string]string, cl client.Client) (*SymmetricRoutingController, error) {
+	// Check the validity of input parameters.
+	if routingTableID > unix.RT_TABLE_MAX {
+		return nil, &liqoerrors.WrongParameter{Parameter: "routingTableID", Reason: liqoerrors.MinorOrEqual + strconv.Itoa(unix.RT_TABLE_MAX)}
+	}
+	if routingTableID < 0 {
+		return nil, &liqoerrors.WrongParameter{Parameter: "routingTableID", Reason: liqoerrors.GreaterOrEqual + strconv.Itoa(0)}
+	}
+	if vxlanDevice.Link == nil {
+		return nil, &liqoerrors.WrongParameter{Parameter: "vxlanDevice.Link", Reason: liqoerrors.NotNil}
+	}
+	return &SymmetricRoutingController{
+		Client:         cl,
+		vxlanDev:       vxlanDevice,
+		nodeName:       nodeName,
+		routingTableID: routingTableID,
+		nodesLock:      nodesLock,
+		vxlanNodes:     vxlanNodes,
+		routes:         map[string]string{},
+	}, nil
+}
+
+// addPeer adds route for the given pod. It return true when the route dose not exists
+// and is added, false if the route does already exist. An error is return if something
+// goes wrong.
+func (src *SymmetricRoutingController) addRoute(req ctrl.Request, p *corev1.Pod) (bool, error) {
```

Comment:
Not sure if the lock is necessary. AFAIK Reconcile is by default executed without any parallelization.

Reply:
The lock is need for the `src.vxlanNodes` map. The map is populated (write access) by another go routine. The operator here accesses it in `read` mode.